File: 214/projects/08/clojure/src/Temperature.clj
;; TemperatureTable.cljis a module for Temperautres
;;
;; Usage: clojure -m TemperatureTable
;;
;; Completed by: Duncan Van Keulen
;; Date: 4/12/2020
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require '[clojure.string :as str]) ;; include the Temperature module

;; Define a Temperature record with degree and scale attributes
(defrecord Temperature [myDegree myScale])

(defn make-Temperature [degree scale]
  (->Temperature degree scale)
)

(defn getDegree [^Temperature temp] 
  (:myDegree temp)
)

(defn getScale [^Temperature temp]
  (:myScale temp)
)

(defn isValidTemperature [degree scale] 
  (cond
    (= \F scale) (if (< degree -460.0) false true)
    (= \f scale) (if (< degree -460.0) false true)
    (= \C scale) (if (< degree -273.15) false true)
    (= \c scale) (if (< degree -273.15) false true)
    (= \K scale) (if (< degree 0.0) false true)
    (= \k scale) (if (< degree 0.0) false true)
    :else false
  )   
)

(defn toFahrenheit [^Temperature temp]
  (cond
    ;; From Fahrenheit
    (= \F (getScale temp)) temp ;; no conversion needed
    (= \f (getScale temp)) temp ;; no conversion needed

    ;; From Celsius
    (= \C (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (getDegree temp)) 32.0) \F)
    (= \c (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (getDegree temp)) 32.0) \F)

    ;; From Kelvin
    (= \K (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (- (getDegree temp) 273.15)) 32.0) \F)
    (= \k (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (- (getDegree temp) 273.15)) 32.0) \F)

    :else false
  )
)

(defn toCelsius [^Temperature temp]
  (cond
    ;; From Fahrenheit
    (= \F (getScale temp)) (make-Temperature (* (/ 5.0 9.0) (- (getDegree temp) 32.0)) \C)
    (= \f (getScale temp)) (make-Temperature (* (/ 5.0 9.0) (- (getDegree temp) 32.0)) \C)

    ;; From Celsius
    (= \C (getScale temp)) temp ;; no conversion needed
    (= \c (getScale temp)) temp ;; no conversion needed

    ;; From Kelvin
    (= \K (getScale temp)) (make-Temperature (- (getDegree temp) 273.15) \C)
    (= \k (getScale temp)) (make-Temperature (- (getDegree temp) 273.15) \C)

    :else false
  )
)

(defn toKelvin [^Temperature temp] 
  (cond
    ;; From Fahrenheit
    (= \F (getScale temp)) (make-Temperature (+ (* (- (getDegree temp) 32.0) (/ 5.0 9.0)) 273.15) \K)
    (= \f (getScale temp)) (make-Temperature (+ (* (- (getDegree temp) 32.0) (/ 5.0 9.0)) 273.15) \K)

    ;; From Celsius
    (= \C (getScale temp)) (make-Temperature (+ (getDegree temp) 273.15) \K)
    (= \c (getScale temp)) (make-Temperature (+ (getDegree temp) 273.15) \K)

    ;; From Kelvin
    (= \K (getScale temp)) temp ;; no conversion needed
    (= \k (getScale temp)) temp ;; no conversion needed

    :else false
  )
)

(defn enterTemperature []
  (let 
    [
      input (read-line)
      splicedInput (str/split input #" ")
      inputDegree (Float/parseFloat (first splicedInput))
      inputScale (get (second splicedInput) 0)
    ]
    (if (isValidTemperature inputDegree inputScale)
      (->Temperature inputDegree inputScale)
      (do (println "Couldn't enter temperature") (flush))
    )
  )
)

(defn displayTemperature [^Temperature temp]
  (print (format "%.5f"(getDegree temp)) " " (getScale temp))
)

(defn raiseTemperature [^Temperature temp degree]
  (if (isValidTemperature (+ (getDegree temp) degree) (getScale temp)) 
    ;; True
    (make-Temperature (+ (getDegree temp) degree) (getScale temp))
    ;; False
    (do (print "Could not raise temperature") (flush))
  )  
)

(defn lowerTemperature [^Temperature temp degree]
  (if (isValidTemperature (- (getDegree temp) degree) (getScale temp)) 
    ;; True
    (make-Temperature (- (getDegree temp) degree) (getScale temp))
    ;; False
    (do (print "Could not lower temperature") (flush))
  )  
)

(defn equals [^Temperature temp1 ^Temperature temp2]
  (if (= (getDegree (toFahrenheit temp1)) (getDegree (toFahrenheit temp2)))
    ;; True-Section
      true ;; return true
    ;; False-Section
      false ;; return false
  )
)

(defn lessThan [^Temperature temp1 ^Temperature temp2]
  (if (< (getDegree (toFahrenheit temp1)) (getDegree (toFahrenheit temp2)))
    ;; True-Section
      true ;; return true
    ;; False-Section
      false ;; return false
  )
)

File: 214/projects/08/ruby/Temperature.rb
# Temperature class module file
# 
# Duncan Van Keulen for CS214 Project08
# 4/13/2020
####################################

module TemperatureModule
    class Temperature

        ###################################################################
        # isValid() checks whether a Temperature type is valid          
        # Receive: degrees, myDegree Float;
        #          scale, myScale Character.         
        # Return: true if Temperature type is valid, false otherwise.     
        ###################################################################
        def isValidTemperature(degree, scale)
            case scale
            when 'F', 'f'
                if degree < -460.0
                    puts "Cannot have temperature below absolute zero (F)"
                    return false
                else
                    return true
                end #if
            when 'C', 'c'
                if degree < -273.15
                    puts "Cannot have temperature below absolute zero (C)"
                    return false
                else
                    return true
                end #if
            when 'K', 'k'
                if degree < 0
                    puts "Cannot have temperature below absolute zero (K)"
                    return false
                else
                    return true
                end #if
            else
                puts "Invalid temperature"
                return false
            end #case
        end #isValidTemperature

        def initialize(degree, scale)
            if isValidTemperature(degree, scale)
                @myDegree, @myScale = degree, scale
            end
        end

        attr_reader :myDegree, :myScale

        def raiseTemperature(degrees)
            if isValidTemperature((self.myDegree + degrees), @myScale)
                return Temperature.new((@myDegree + degrees), @myScale)
            else
                puts "Could not raise temperature"
            end #if
        end #raiseTemperature

        def lowerTemperature(degrees)
            if isValidTemperature((@myDegrees - degrees), @myScale)
                return Temperature.new((@myDegree - degrees), @myScale)
            else
                puts "Could not raise temperature"
            end #if
        end #lowerTemperature

        def toFahrenheit()
            case @myScale
            when 'F', 'f'
                # Already in Fahrenheit
                return self
            when 'C', 'c'
                newDeg = ((9.0/5.0) * @myDegree) + 32.0
                return Temperature.new(newDeg, 'F')
            when 'K', 'k'
                newDeg = ((@myDegree - 273.15) * (9.0/5.0)) + 32.0
                return Temperature.new(newDeg, 'F')
            else
                puts "Invalid temperature scale [toF]"
                return false
            end #case
        end #toFahrenheit

        def toCelsius()
            case @myScale
            when 'F', 'f'
                newDeg = (5.0/9.0) * (@myDegree - 32.0)
                return Temperature.new(newDeg, 'C')
            when 'C', 'c'
                # Already in Celsius
                return self
            when 'K', 'k'
                newDeg = @myDegree - 273.15
                return Temperature.new(newDeg, 'C')
            else
                puts "Invalid temperature scale [toC]"
                return false
            end #case
        end #toCelsius

        def toKelvin()
            case @myScale
            when 'F', 'f'
                newDeg = ((@myDegree - 32) * (5.0 / 9.0)) + 273.15
                return Temperature.new(newDeg, 'K')
            when 'C', 'c'
                newDeg = @myDegree + 273.15
                return Temperature.new(newDeg, 'K')
            when 'K', 'k'
                # Already in Kevlin
                return self
            else
                puts "Invalid temperature [toK]"
                return false
            end #case
        end #toFahrenheit

        def enterTemperature()
            input = gets.split
            degree = input[0].to_f
            scale = input[1].to_s

            return Temperature.new(degree, scale)
        end

        def displayTemperature()
            print '%.5f' % @myDegree + " " + @myScale + "\t"
        end

        def equals(temp)
            return self.toFahrenheit().myDegree == temp.toFahrenheit.myDegree
        end

        def lessThan(temp)
            return self.toFahrenheit().myDegree < temp.toFahrenheit.myDegree
        end

    end #class Temperature
end #module TemperatureModule
